debezium-core/src/main/java/io/debezium/util/OrderedIdBuilder.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.util;
+
+import org.apache.kafka.connect.data.Schema;
+
+public abstract class OrderedIdBuilder {
```

Comment:
How about this contract:

```
public abstract class OrderedIdBuilder {
    public void addOrderedId(Struct sourceInfo);
    public void restoreState(String lastState);
    public OrderedIdBuilder clone(); // not quite sure on that one yet
}
```

That'd move most code into implementations of this contract themselves (e.g. the no-op handling would be fully encapsulated), simplifying the code in the connectors (whose `SourceInfo#struct()` method just would have to call `addOrderedId()` whenever invoked).

Reply:
As far as I can tell, all the implementations of SourceInfo look more like this:

1. update the position in the log (and how I have implemented it, bump the order_id)
2. build the current position as a `Map<String, Object>`  to store in the offsetStore
3. build the struct that is actually written out in a message

So I think there are two issues with that API:
1. The struct and the map are both required, especially for the counter,  so we would at least need a method to get the current state so we can also generate the map that is written to the offsetStore.
2. Because the calls happen in the order described above, if you don't get the next id until you actually build the struct, then the map you build will be *before* you generate the new ID, meaning there will be an off by one error. 

I think a simpler API could exist, but it seems like it would need to change with SourceInfo API to be more controlled for that to work.